---
title: 'Empowering the community with notebooks for bespoke microbiome analyses'
title_short: 'BioHackEU22 #12: MGnify Notebooks'
tags:
  - metagenomics
  - jupyter
  - galaxy project
authors:
  - name: Alexander B Rogers
  - orcid: 0000-0002-4283-6135
    affiliation: 1
  - name: Martin Beracochea
    orcid: 0000-0003-3472-3736
    affiliation: 1
affiliations:
  - name: EMBL-EBI
    index: 1
  - name: Second Affiliation
    index: 2
date: 10 November 2022
cito-bibliography: paper.bib
event: BH22EU
biohackathon_name: "BioHackathon Europe 2022"
biohackathon_url:   "https://biohackathon-europe.org/"
biohackathon_location: "Paris, France, 2022"
group: Project 12
# URL to project git repo --- should contain the actual paper.md:
git_url: https://github.com/EBI-Metagenomics/mgnify-biohackeu22-report
# This is the short authors description that is used at the
# bottom of the generated paper (typically the first two authors):
authors_short: Rogers, Beracochea, \emph{et al.}
---


# Introduction

As part of the BioHackathon Europe 2022, we introduced several developments to the "MGnify Notebook Server"
ranging from running on new infrastructure to adding new content and documentation.

MGnify is EMBL-EBI’s metagenomics resource, which is part of ELIXIR Metagenomics Community. 
In the last year, MGnify launched a Notebook Server to provide an online, Jupyter Lab environment for users to explore programmatic access to MGnify’s datasets using Python or with R. 
This ready to use environment and example analysis notebooks bridge the gap between the ease but limitations of browsing the MGnify website, and the complexity but possibilities of installing a local environment to work with data stored in MGnify. 
Particular goals of the Notebook Server include reproducible downstream analyses, user empowerment through best-practice examples and fast workflows from datasets to publication-ready graphics, and code-as-documentation training materials for users of MGnify.

As a young resource, there were potential improvements to be made across the entire technology and content stack.

# Results

## MGnify notebooks running on Galaxy infrastructure

## Enhanced documentation and clearer user journeys

## Improving interoperability of IGV as a Jupyter Lab Widget

## Integration between MGnify's Protein Database and ESM Atlas

## Notebooks covering MGnify's metagenome-assembled genomes

# Discussion

...

## Acknowledgements

...

## References
